File: code-trimmer-python/codetrimmer/service/report_generator.py
"""Report generation service."""

import csv
import io
import json
import logging
import sqlite3
from pathlib import Path

from codetrimmer.error.codes import ErrorCode
from codetrimmer.error.exceptions import CodeTrimmerException
from codetrimmer.model.statistics import ProcessingStatistics

logger = logging.getLogger(__name__)


class ReportGenerator:
    """Service for generating processing reports.

    Supports JSON, CSV, and SQLite output formats.
    """

    def export_json(self, stats: ProcessingStatistics) -> str:
        """Export statistics to JSON format.

        Args:
            stats: Processing statistics

        Returns:
            JSON string
        """
        return json.dumps(stats.to_dict(), indent=2)

    def export_csv(self, stats: ProcessingStatistics) -> str:
        """Export statistics to CSV format.

        Args:
            stats: Processing statistics

        Returns:
            CSV string
        """
        output = io.StringIO()
        writer = csv.writer(output)

        # Write header
        writer.writerow([
            "file_path",
            "was_modified",
            "bytes_modified",
            "error_code",
            "error_message",
            "backup_path",
            "processing_time_ms",
        ])

        # Write data rows
        for result in stats.results:
            writer.writerow([
                result.file_path,
                result.was_modified,
                result.bytes_modified,
                result.error.code if result.error else "",
                result.error_message,
                result.backup_path or "",
                result.processing_time_ms,
            ])

        return output.getvalue()

    def export_sqlite(
        self,
        stats: ProcessingStatistics,
        db_path: str,
    ) -> None:
        """Export statistics to SQLite database.

        Args:
            stats: Processing statistics
            db_path: Path to SQLite database file

        Raises:
            CodeTrimmerException: If database operation fails
        """
        try:
            conn = sqlite3.connect(db_path)
            cursor = conn.cursor()

            # Create tables
            cursor.execute("""
                CREATE TABLE IF NOT EXISTS processing_runs (
                    id INTEGER PRIMARY KEY AUTOINCREMENT,
                    timestamp DATETIME DEFAULT CURRENT_TIMESTAMP,
                    files_processed INTEGER,
                    files_modified INTEGER,
                    files_skipped INTEGER,
                    files_with_errors INTEGER,
                    total_bytes_trimmed INTEGER,
                    total_execution_time_ms INTEGER
                )
            """)

            cursor.execute("""
                CREATE TABLE IF NOT EXISTS file_results (
                    id INTEGER PRIMARY KEY AUTOINCREMENT,
                    run_id INTEGER,
                    file_path TEXT,
                    was_modified INTEGER,
                    bytes_modified INTEGER,
                    error_code TEXT,
                    error_message TEXT,
                    backup_path TEXT,
                    processing_time_ms INTEGER,
                    FOREIGN KEY (run_id) REFERENCES processing_runs (id)
                )
            """)

            # Insert run summary
            cursor.execute("""
                INSERT INTO processing_runs (
                    files_processed, files_modified, files_skipped,
                    files_with_errors, total_bytes_trimmed, total_execution_time_ms
                ) VALUES (?, ?, ?, ?, ?, ?)
            """, (
                stats.files_processed,
                stats.files_modified,
                stats.files_skipped,
                stats.files_with_errors,
                stats.total_bytes_trimmed,
                stats.total_execution_time_ms,
            ))

            run_id = cursor.lastrowid

            # Insert file results
            for result in stats.results:
                cursor.execute("""
                    INSERT INTO file_results (
                        run_id, file_path, was_modified, bytes_modified,
                        error_code, error_message, backup_path, processing_time_ms
                    ) VALUES (?, ?, ?, ?, ?, ?, ?, ?)
                """, (
                    run_id,
                    result.file_path,
                    1 if result.was_modified else 0,
                    result.bytes_modified,
                    result.error.code if result.error else None,
                    result.error_message,
                    result.backup_path,
                    result.processing_time_ms,
                ))

            conn.commit()
            conn.close()

            logger.info("Report exported to SQLite: %s", db_path)

        except sqlite3.Error as e:
            raise CodeTrimmerException(
                ErrorCode.CT_0063,
                f"SQLite error: {e}",
                "Check database path and permissions",
            ) from e

    def save_report(
        self,
        stats: ProcessingStatistics,
        output_path: str,
        format: str = "json",
    ) -> None:
        """Save report to file.

        Args:
            stats: Processing statistics
            output_path: Path to output file
            format: Report format (json, csv, sqlite)

        Raises:
            CodeTrimmerException: If report generation fails
        """
        try:
            if format == "json":
                content = self.export_json(stats)
                Path(output_path).write_text(content, encoding="utf-8")
            elif format == "csv":
                content = self.export_csv(stats)
                Path(output_path).write_text(content, encoding="utf-8")
            elif format == "sqlite":
                self.export_sqlite(stats, output_path)
            else:
                raise CodeTrimmerException(
                    ErrorCode.CT_0061,
                    f"Invalid report format: {format}",
                    "Use json, csv, or sqlite",
                )

            logger.info("Report saved to: %s", output_path)

        except IOError as e:
            raise CodeTrimmerException(
                ErrorCode.CT_0060,
                f"Failed to save report: {e}",
                "Check disk space and permissions",
            ) from e

    async def send_to_endpoint(
        self,
        stats: ProcessingStatistics,
        url: str,
    ) -> None:
        """Send report to HTTP endpoint.

        Args:
            stats: Processing statistics
            url: HTTP endpoint URL

        Raises:
            CodeTrimmerException: If sending fails
        """
        try:
            import aiohttp

            async with aiohttp.ClientSession() as session:
                async with session.post(
                    url,
                    json=stats.to_dict(),
                    headers={"Content-Type": "application/json"},
                ) as response:
                    if response.status >= 400:
                        raise CodeTrimmerException(
                            ErrorCode.CT_0062,
                            f"HTTP error {response.status}",
                            "Check endpoint URL and availability",
                        )

            logger.info("Report sent to endpoint: %s", url)

        except ImportError:
            raise CodeTrimmerException(
                ErrorCode.CT_0062,
                "aiohttp not installed",
                "Install with: pip install codetrimmer[http]",
            )
        except Exception as e:
            raise CodeTrimmerException(
                ErrorCode.CT_0062,
                f"Failed to send report: {e}",
                "Check endpoint URL and network connection",
            ) from e

File: src/main/java/com/codetrimmer/report/ReportGenerator.java
package com.codetrimmer.report;

import com.codetrimmer.error.CodeTrimmerException;
import com.codetrimmer.error.ErrorCode;
import com.codetrimmer.model.ProcessingStatistics;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.SerializationFeature;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;

import java.io.IOException;
import java.io.StringWriter;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.PreparedStatement;
import java.sql.SQLException;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.LinkedHashMap;
import java.util.Map;

/**
 * Service for generating and exporting processing reports.
 * Supports JSON, CSV, and SQLite formats.
 */
@Service
public class ReportGenerator {

    private static final Logger LOGGER = LoggerFactory.getLogger(ReportGenerator.class);
    private final ObjectMapper objectMapper;

    public ReportGenerator() {
        this.objectMapper = new ObjectMapper();
        this.objectMapper.enable(SerializationFeature.INDENT_OUTPUT);
    }

    /**
     * Generates a report in the specified format.
     *
     * @param stats the processing statistics
     * @param format the output format (json, csv, sqlite)
     * @param outputPath the output file path
     */
    public void generateReport(ProcessingStatistics stats, String format, String outputPath) {
        if (format == null || format.trim().isEmpty()) {
            return;
        }

        String normalizedFormat = format.toLowerCase().trim();
        try {
            switch (normalizedFormat) {
                case "json" -> generateJsonReport(stats, outputPath);
                case "csv" -> generateCsvReport(stats, outputPath);
                case "sqlite" -> generateSqliteReport(stats, outputPath);
                default -> throw new CodeTrimmerException(
                    ErrorCode.CT_0061,
                    "Unknown format: " + format,
                    "Use json, csv, or sqlite"
                );
            }
            LOGGER.info("Report generated: {}", outputPath);
        } catch (CodeTrimmerException e) {
            throw e;
        } catch (Exception e) {
            throw new CodeTrimmerException(ErrorCode.CT_0060, e);
        }
    }

    /**
     * Generates a JSON report.
     *
     * @param stats the processing statistics
     * @param outputPath the output file path
     */
    public void generateJsonReport(ProcessingStatistics stats, String outputPath) throws IOException {
        Map<String, Object> reportData = createReportData(stats);
        String json = objectMapper.writeValueAsString(reportData);

        if (outputPath != null && !outputPath.trim().isEmpty()) {
            Files.writeString(Paths.get(outputPath), json);
        } else {
            System.out.println(json);
        }
    }

    /**
     * Gets JSON report as a string.
     *
     * @param stats the processing statistics
     * @return JSON string
     */
    public String getJsonReport(ProcessingStatistics stats) {
        try {
            Map<String, Object> reportData = createReportData(stats);
            return objectMapper.writeValueAsString(reportData);
        } catch (Exception e) {
            throw new CodeTrimmerException(ErrorCode.CT_0060, e);
        }
    }

    /**
     * Generates a CSV report.
     *
     * @param stats the processing statistics
     * @param outputPath the output file path
     */
    public void generateCsvReport(ProcessingStatistics stats, String outputPath) throws IOException {
        StringWriter writer = new StringWriter();

        // CSV Header
        String header = "timestamp,files_scanned,files_modified,files_skipped,"
            + "lines_trimmed,blank_lines_removed,execution_time_ms\n";

        // CSV Data
        writer.write(LocalDateTime.now().format(DateTimeFormatter.ISO_LOCAL_DATE_TIME));
        writer.write(",");
        writer.write(String.valueOf(stats.getFilesScanned()));
        writer.write(",");
        writer.write(String.valueOf(stats.getFilesModified()));
        writer.write(",");
        writer.write(String.valueOf(stats.getFilesSkipped()));
        writer.write(",");
        writer.write(String.valueOf(stats.getLinesTrimmed()));
        writer.write(",");
        writer.write(String.valueOf(stats.getBlankLinesRemoved()));
        writer.write(",");
        writer.write(String.valueOf(stats.getExecutionTimeMs()));
        writer.write("\n");

        String dataLine = writer.toString();

        if (outputPath != null && !outputPath.trim().isEmpty()) {
            Path path = Paths.get(outputPath);
            if (Files.exists(path)) {
                // Append to existing file (without header)
                Files.write(path, dataLine.getBytes(java.nio.charset.StandardCharsets.UTF_8),
                    java.nio.file.StandardOpenOption.APPEND);
            } else {
                Files.writeString(path, header + dataLine);
            }
        } else {
            System.out.println(header + dataLine);
        }
    }

    /**
     * Gets CSV report as a string.
     *
     * @param stats the processing statistics
     * @return CSV string
     */
    public String getCsvReport(ProcessingStatistics stats) {
        StringWriter writer = new StringWriter();
        writer.write("timestamp,files_scanned,files_modified,files_skipped,");
        writer.write("lines_trimmed,blank_lines_removed,execution_time_ms\n");
        writer.write(LocalDateTime.now().format(DateTimeFormatter.ISO_LOCAL_DATE_TIME));
        writer.write(",");
        writer.write(String.valueOf(stats.getFilesScanned()));
        writer.write(",");
        writer.write(String.valueOf(stats.getFilesModified()));
        writer.write(",");
        writer.write(String.valueOf(stats.getFilesSkipped()));
        writer.write(",");
        writer.write(String.valueOf(stats.getLinesTrimmed()));
        writer.write(",");
        writer.write(String.valueOf(stats.getBlankLinesRemoved()));
        writer.write(",");
        writer.write(String.valueOf(stats.getExecutionTimeMs()));
        writer.write("\n");
        return writer.toString();
    }

    /**
     * Generates a SQLite report.
     *
     * @param stats the processing statistics
     * @param outputPath the SQLite database path
     */
    public void generateSqliteReport(ProcessingStatistics stats, String outputPath) {
        String dbPath = outputPath != null && !outputPath.trim().isEmpty()
            ? outputPath : "codetrimmer_stats.db";

        String url = "jdbc:sqlite:" + dbPath;

        try (Connection conn = DriverManager.getConnection(url)) {
            createTableIfNotExists(conn);
            insertStats(conn, stats);
            LOGGER.info("Statistics saved to SQLite database: {}", dbPath);
        } catch (SQLException e) {
            throw new CodeTrimmerException(
                ErrorCode.CT_0063,
                "Failed to write to SQLite: " + e.getMessage(),
                "Check database permissions and disk space"
            );
        }
    }

    private void createTableIfNotExists(Connection conn) throws SQLException {
        String sql = """
            CREATE TABLE IF NOT EXISTS processing_stats (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                timestamp TEXT NOT NULL,
                files_scanned INTEGER,
                files_modified INTEGER,
                files_skipped INTEGER,
                lines_trimmed INTEGER,
                blank_lines_removed INTEGER,
                execution_time_ms INTEGER
            )
            """;
        conn.createStatement().execute(sql);
    }

    private void insertStats(Connection conn, ProcessingStatistics stats) throws SQLException {
        String sql = """
            INSERT INTO processing_stats
            (timestamp, files_scanned, files_modified, files_skipped,
             lines_trimmed, blank_lines_removed, execution_time_ms)
            VALUES (?, ?, ?, ?, ?, ?, ?)
            """;

        try (PreparedStatement pstmt = conn.prepareStatement(sql)) {
            pstmt.setString(1, LocalDateTime.now().format(DateTimeFormatter.ISO_LOCAL_DATE_TIME));
            pstmt.setInt(2, stats.getFilesScanned());
            pstmt.setInt(3, stats.getFilesModified());
            pstmt.setInt(4, stats.getFilesSkipped());
            pstmt.setInt(5, stats.getLinesTrimmed());
            pstmt.setInt(6, stats.getBlankLinesRemoved());
            pstmt.setLong(7, stats.getExecutionTimeMs());
            pstmt.executeUpdate();
        }
    }

    private Map<String, Object> createReportData(ProcessingStatistics stats) {
        Map<String, Object> data = new LinkedHashMap<>();
        data.put("timestamp", LocalDateTime.now().format(DateTimeFormatter.ISO_LOCAL_DATE_TIME));
        data.put("filesScanned", stats.getFilesScanned());
        data.put("filesModified", stats.getFilesModified());
        data.put("filesSkipped", stats.getFilesSkipped());
        data.put("linesTrimmed", stats.getLinesTrimmed());
        data.put("blankLinesRemoved", stats.getBlankLinesRemoved());
        data.put("executionTimeMs", stats.getExecutionTimeMs());
        data.put("executionTimeSec", stats.getExecutionTimeSec());
        return data;
    }

    /**
     * Sends report to HTTP endpoint.
     *
     * @param stats the processing statistics
     * @param endpoint the HTTP endpoint URL
     */
    public void sendToEndpoint(ProcessingStatistics stats, String endpoint) {
        if (endpoint == null || endpoint.trim().isEmpty()) {
            return;
        }

        try {
            String json = getJsonReport(stats);
            java.net.http.HttpClient client = java.net.http.HttpClient.newHttpClient();
            java.net.http.HttpRequest request = java.net.http.HttpRequest.newBuilder()
                .uri(java.net.URI.create(endpoint))
                .header("Content-Type", "application/json")
                .POST(java.net.http.HttpRequest.BodyPublishers.ofString(json))
                .build();

            java.net.http.HttpResponse<String> response = client.send(request,
                java.net.http.HttpResponse.BodyHandlers.ofString());

            if (response.statusCode() >= 200 && response.statusCode() < 300) {
                LOGGER.info("Report sent to endpoint: {}", endpoint);
            } else {
                throw new CodeTrimmerException(
                    ErrorCode.CT_0062,
                    "HTTP error " + response.statusCode(),
                    "Check endpoint URL and availability"
                );
            }
        } catch (CodeTrimmerException e) {
            throw e;
        } catch (Exception e) {
            throw new CodeTrimmerException(ErrorCode.CT_0062, e);
        }
    }
}
